### DISCUSSION TITLE
AtomWithRefresh Exstension to support other use cases

### DISCUSSION THREAD

**Discussion by dvd900** [pmndrs/jotai#2278] [General] [answered]
I need to be able to refresh an atom without triggering a suspension (think background update/poll). I'm aware of the loadable util as well but I really like the ease of the `atomWithRefresh` 

Here's where I landed, until I found out that atoms are only writable if an initial value is provided. 

```
function atomWithKeepRefresh<T>(fn: (get: Getter) => T) {
  const refreshCounter = atom(0);
  const valueAtom = atom(
    (get) => {
      get(refreshCounter);
      return fn(get);
    }, (_, set, arg) => {
      set(valueAtom, arg)
    }
  )

  return atom<T, [boolean?], unknown>(
    (get) => {
      return get(valueAtom);
    },
    async (get, set, keep) => {
      if(keep) {
        const response = await fn(get);
        set(valueAtom, response)
      } else {
        set(refreshCounter, (i) => i + 1)
      }
    },
  );
}
```

So this leaves me with the question, what's a good way to approach this?

**Comment by dai-shi** [comment]
If you don't need suspending, I don't think you need `atomWithRefresh`. Instead just write a resolved value to an atom, as you did with `keep=true`. btw, you need one more atom to make it writeable or use `atomWithDefault`.

If you want to use `atomWithRefresh` and selectively avoid suspending. `unwrap` should work.

**Reply by dvd900** [reply]
yes I use suspension, I don't think unwrap works for this because I want the selective suspension to determined by the component calling refresh(true)

**Reply by dvd900** [reply]
```
export function atomWithRefresh<T>(fn: (get: Getter) => T) {
  const storeAtom = atomWithDefault((get) => {
    return fn(get);
  });
  
  return atom<T, [boolean?], unknown>(
    (get) => get(storeAtom),
    async (get, set, keep) => {
      if (keep) {
        const response = await fn(get);
        set(storeAtom, response);
      } else {
        set(storeAtom, fn(get));
      }
    }
  );
}
```

using atomWithDefault causes the default (a async request) to be run over and over again.

**Comment by dai-shi** [comment]
> yes I use suspension, I don't think unwrap works for this because I want the selective suspension to determined by the component calling refresh(true)

That sounds like a complicated scenario, but still can be done to selectively use `unwrap`ped one or original one.

**Reply by dvd900** [reply]
think about it like a background update. I have some data source that determines the flavor of the application, along with some other stuff. On the initial load or refreshes I want to suspend while I wait for it to resolve. On the other hand there exists some other effects (puts to api's) in the application may change it, but optimistically i prefer to keep the original value and carry on.

**Reply by dai-shi** [reply, marked as answer]
that explains why unwrap isn't best fit.

```ts
const fetchData = async () => { /* ... */ };
const baseDataAtom = atom<Data | Promise<Data> | null>(null);
const dataAtom = atom((get) => get(baseDataAtom)) ?? fetchData());
const refetchDataAtom = atom(null, (get, set, suspend) => {
  const promise = fetchData();
  if (suspend) {
    set(baseDataAtom, promise)
  } else {
    promise.then((data) => set(baseDataAtom, data));
  }
});
```

See also: https://blog.axlight.com/posts/how-to-use-jotai-and-use-transition-for-mutation/

**Reply by dvd900** [reply]
This is a good base example of what i was trying to do thank you. There ended up being a bug in my application that was conflicting with my understanding of jotai. mounting suspense above the provider caused crazy issues X___X

still i learned a lot thanks 🍻
